Switch quantized_decomposed over to new custom ops API

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123454

We are taking API feedback. Changes:
- I removed some of the default values (they weren't being used).
- I was unable to convert the last op (which is essentially an
  autograd.Function registered as CompositeImplicitAutograd). That one
  is "incorrectly registered"; I punt fixing it to the future.

Test Plan:
- existing tests

Differential Revision: [D56151990](https://our.internmc.facebook.com/intern/diff/D56151990)

zou3519 committed:
Switch quantized_decomposed over to new custom ops API

We are taking API feedback. Changes:
- I removed some of the default values (they weren't being used).
- I was unable to convert the last op (which is essentially an
  autograd.Function registered as CompositeImplicitAutograd). That one
  is "incorrectly registered"; I punt fixing it to the future.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Switch quantized_decomposed over to new custom ops API"


We are taking API feedback. Changes:
- I removed some of the default values (they weren't being used).
- I was unable to convert the last op (which is essentially an
  autograd.Function registered as CompositeImplicitAutograd). That one
  is "incorrectly registered"; I punt fixing it to the future.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Switch quantized_decomposed over to new custom ops API"


We are taking API feedback. Changes:
- I removed some of the default values (they weren't being used).
- I was unable to convert the last op (which is essentially an
  autograd.Function registered as CompositeImplicitAutograd). That one
  is "incorrectly registered"; I punt fixing it to the future.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Switch quantized_decomposed over to new custom ops API"


We are taking API feedback. Changes:
- I removed some of the default values (they weren't being used).
- I was unable to convert the last op (which is essentially an
  autograd.Function registered as CompositeImplicitAutograd). That one
  is "incorrectly registered"; I punt fixing it to the future.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Switch quantized_decomposed over to new custom ops API"


We are taking API feedback. Changes:
- I removed some of the default values (they weren't being used).
- I was unable to convert the last op (which is essentially an
  autograd.Function registered as CompositeImplicitAutograd). That one
  is "incorrectly registered"; I punt fixing it to the future.

Test Plan:
- existing tests

[ghstack-poisoned]

albanD reviewed (commented):
Pretty cool!
Did you check that all the schemas are exactly the same including kwargs-only, etc?

andrewor14 commented on torch/ao/quantization/fx/_decomposed.py:
```diff
@@ -560,8 +521,8 @@ def dequantize_per_channel(
-@impl(quantized_decomposed_lib, "dequantize_per_channel", "Meta")
-def dequantize_per_channel_meta(
+@dequantize_per_channel.register_fake
+def _(
```
If I have multiple of these in the same file will it mess with the namespace?

zou3519 replied:
Nope, the `_` is just like any other name in python (e.g. `impl`)

andrewor14 commented on torch/ao/quantization/fx/_decomposed.py:
```diff
@@ -824,25 +747,16 @@ def dequantize_per_token_meta(
-@impl(
-    quantized_decomposed_lib, "quantize_per_channel_group", "CompositeExplicitAutograd"
-)
+@custom_op(f"{ns}::quantize_per_channel_group", mutates_args=())
```
Seems like `mutates_args` is mostly empty (100% of the case in this file). Did we decide to make this a required arg or can we make it optional?

andrewor14 replied:
Also, have you considered letting the user set an explicit flag that says "get the op name from the function name" to skip the duplicate naming? So if the user sets this to true then they're acknowledging the risk of other decorators messing the op name up. Maybe something like:
```
# turns into `quantized_decomposed::quantize_per_channel_group`
@custom_op(ns, append_func_name_to_op_name=True)
def quantize_per_channel_group(...)
```
Do you think this idea makes sense or it's still too risky?

zou3519 replied:
> Seems like mutates_args is mostly empty (100% of the case in this file). Did we decide to make this a required arg or can we make it optional?

Yes, we decided to make this required. If the user does not tell us correctly what args are being mutated, this will lead to undefined behavior. If it is not required, our experience with the previous custom ops API is that people will copy-paste existing custom ops and never specify it.

> Also, have you considered letting the user set an explicit flag that says "get the op name from the function name" to skip the duplicate naming? So if the user sets this to true then they're acknowledging the risk of other decorators messing the op name up. Maybe something like:

The explicit name seemed easier to grok. But also something else is that with overload names (e.g. "quantize_decomposed::quantize_for_tensor.tensor") there is no way to infer this from the function's name.

andrewor14 commented on torch/ao/quantization/fx/_decomposed.py:
```diff
@@ -908,27 +822,17 @@ def quantize_per_channel_group_meta(
-    output_dtype: torch.dtype = torch.float32,
-):
+    group_size: int,
+    output_dtype: torch.dtype,
```
We can't have default args anymore?

zou3519 replied:
None of the default args I removed were actually being used: they weren't defined in the operator's schema string

andrewor14 commented on torch/ao/quantization/fx/_decomposed.py:
```diff
@@ -965,6 +869,8 @@ def dequantize_per_channel_group(
+
 quantized_decomposed_lib.define(
     "fake_quant_per_channel(Tensor input, Tensor scales, Tensor zero_points, int axis, "
     "int quant_min, int quant_max) -> Tensor")
```
Should we add a TODO for this one to migrate it too? Maybe explain in the comment why we can't do this now?

zou3519 replied:
Yes, I'll add a TODO

andrewor14 reviewed (approved):
Thanks!

zou3519 committed:
Update on "Switch quantized_decomposed over to new custom ops API"


We are taking API feedback. Changes:
- I removed some of the default values (they weren't being used).
- I was unable to convert the last op (which is essentially an
  autograd.Function registered as CompositeImplicitAutograd). That one
  is "incorrectly registered"; I punt fixing it to the future.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge -f "unrelated failure"

DanilBaibak commented:
@pytorchbot revert -m "Break internal build" -c ghfirst

DanilBaibak commented:
Hi @zou3519! Sorry, I need to revert your PR because it breaks the internal build. Here you can find more information about the issue -  [D56057957](https://www.internalfb.com/diff/D56057957)

zou3519 committed:
Update on "Switch quantized_decomposed over to new custom ops API"


We are taking API feedback. Changes:
- I removed some of the default values (they weren't being used).
- I was unable to convert the last op (which is essentially an
  autograd.Function registered as CompositeImplicitAutograd). That one
  is "incorrectly registered"; I punt fixing it to the future.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56151990).
